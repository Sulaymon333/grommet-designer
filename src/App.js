import React from 'react';
import ReactGA from 'react-ga';
import { Grommet, Grid, Keyboard, ResponsiveContext, grommet } from 'grommet';
import ErrorCatcher from './ErrorCatcher';
import Canvas from './Canvas';
import Properties from './Properties/Properties';
import Tree from './Tree/Tree';
import {
  getInitialSelected,
  getScreenByPath,
  setupDesign,
  loading,
  publish,
} from './design';
import ScreenDetails from './Properties/ScreenDetails';
import designerLibrary from './libraries/designer';
import grommetLibrary from './libraries/grommet';
import { loadDesign, loadImports, loadTheme } from './design/load';

const designerTheme = {
  ...grommet,
  global: {
    ...grommet.global,
    colors: { background: { dark: '#282828', light: '#f8f8f8' } },
  },
};

const defaultImports = [
  { name: grommetLibrary.name, library: grommetLibrary },
  { name: designerLibrary.name, library: designerLibrary },
];

const getParams = () => {
  const { location } = window;
  const params = {};
  location.search
    .slice(1)
    .split('&')
    .forEach(p => {
      const [k, v] = p.split('=');
      params[k] = decodeURIComponent(v);
    });
  return params;
};

const App = () => {
  const responsive = React.useContext(ResponsiveContext);
  // In the worst case, we need to load a published design, a published theme,
  // and any imports. We load theme as we go into the 'load'
  // state. When they're all ready, we update all of the necessary states.
  const [load, setLoad] = React.useState({});
  const [design, setDesign] = React.useState(setupDesign(loading));
  const [selected, setSelected] = React.useState(getInitialSelected(design));
  const [imports, setImports] = React.useState(defaultImports);
  const libraries = React.useMemo(
    () => imports.filter(i => i.library).map(i => i.library),
    [imports],
  );
  const [theme, setTheme] = React.useState(grommet);
  const [colorMode, setColorMode] = React.useState('dark');
  const [rtl, setRTL] = React.useState();
  const [preview, setPreview] = React.useState(true);
  const [changes, setChanges] = React.useState([]);
  const [changeIndex, setChangeIndex] = React.useState();
  const [designs, setDesigns] = React.useState([]);
  const selectedComponent = React.useMemo(
    () =>
      selected.component ? design.components[selected.component] : undefined,
    [design, selected],
  );

  // initialize analytics
  React.useEffect(() => {
    if (window.location.host !== 'localhost') {
      const {
        location: { pathname },
      } = window;
      ReactGA.initialize('UA-99690204-4');
      ReactGA.pageview(pathname);
    }
  }, []);

  // load initial design
  React.useEffect(() => {
    const {
      location: { pathname },
    } = window;
    const params = getParams();
    loadDesign(
      pathname === '/_new' ? '_new' : params.id,
      nextDesign => {
        let nextSelected;
        if (params.id) {
          const screen =
            (pathname && getScreenByPath(nextDesign, pathname)) ||
            nextDesign.screenOrder[0];
          nextSelected = { screen };
        } else {
          const stored = localStorage.getItem('selected');
          if (stored) nextSelected = JSON.parse(stored);
          else {
            const screen =
              pathname && pathname !== '/_new'
                ? getScreenByPath(nextDesign, pathname)
                : nextDesign.screenOrder[0];
            const component = nextDesign.screens[screen].root;
            nextSelected = { screen, component };
          }
        }
        setLoad(prevLoad => ({
          ...prevLoad,
          design: nextDesign,
          selected: nextSelected,
        }));

        loadTheme(nextDesign.theme, nextTheme => {
          setLoad(prevLoad => ({ ...prevLoad, theme: nextTheme }));
        });

        let loadingImports = [...defaultImports, ...(nextDesign.imports || [])];
        setLoad(prevLoad => ({ ...prevLoad, imports: loadingImports }));
        loadImports(loadingImports, f => {
          loadingImports = f(loadingImports);
          setLoad(prevLoad => ({ ...prevLoad, imports: loadingImports }));
        });
      },
      true,
    );
  }, []);

  // finish loading
  React.useEffect(() => {
    if (
      load &&
      load.design &&
      load.selected &&
      load.theme &&
      load.imports &&
      !load.imports.some(i => i.url && !(i.design || i.library))
    ) {
      const params = getParams();
      setImports(load.imports);
      setTheme(load.theme);
      setDesign(load.design);
      setSelected(load.selected);
      setPreview(params.preview ? JSON.parse(params.preview) : !!params.id);
      setChanges([{ design: load.design, selected: load.selected }]);
      setChangeIndex(0);
      setLoad(undefined);
    }
  }, [load]);

  React.useEffect(() => {
    const stored = localStorage.getItem('designs');
    if (stored) setDesigns(JSON.parse(stored));
  }, []);

  React.useEffect(() => {
    const stored = localStorage.getItem('colorMode');
    if (stored) setColorMode(stored);
    else if (window.matchMedia) {
      setColorMode(
        window.matchMedia('(prefers-color-scheme: dark)').matches
          ? 'dark'
          : 'light',
      );
    }
  }, []);

  React.useEffect(() => {
    const stored = localStorage.getItem('preview');
    if (stored) setPreview(JSON.parse(stored));
  }, []);

  // browser navigation

  // react when user uses browser back and forward buttons
  React.useEffect(() => {
    const onPopState = () => {
      const {
        location: { pathname },
      } = document;
      const screen = getScreenByPath(design, pathname);
      if (screen)
        setSelected({ screen, component: design.screens[screen].root });
    };

    window.addEventListener('popstate', onPopState);
    return () => window.removeEventListener('popstate', onPopState);
  }, [design]);

  // push state when the user navigates
  React.useEffect(() => {
    if (!load) {
      const {
        location: { pathname },
      } = document;
      // track selected screen in browser location, so browser
      // backward/forward controls work
      const screen = design.screens[selected.screen];
      if (screen && screen.path !== pathname) {
        window.history.pushState(undefined, undefined, screen.path);
      }
    }
  }, [design, load, selected.screen]);

  // store design
  React.useEffect(() => {
    if (!load) {
      // do this stuff lazily, so we don't bog down the UI
      const timer = setTimeout(() => {
        document.title = design.name;
        const date = new Date();
        date.setMilliseconds(0);
        design.date = date.toISOString();

        localStorage.setItem(design.name, JSON.stringify(design));
        localStorage.setItem('activeDesign', design.name);

        if (!designs.includes(design.name)) {
          const nextDesigns = [design.name, ...designs];
          localStorage.setItem('designs', JSON.stringify(nextDesigns));
          setDesigns(nextDesigns);
        }
      }, 1000);
      return () => clearTimeout(timer);
    }
  }, [design, designs, load]);

  // align imports with design.imports
  React.useEffect(() => {
    if (!load) {
      // remove any imports we don't want anymore
      const nextImports = imports.filter(
        ({ url }) =>
          !url || design.imports.findIndex(i => i.url === url) !== -1,
      );
      let changed = nextImports.length !== imports.length;
      // add any imports we don't have yet
      design.imports.forEach(({ url }) => {
        if (nextImports.findIndex(i => i.url === url) === -1) {
          nextImports.push({ url });
          changed = true;
        }
      });
      if (changed) setImports(nextImports);
    }
  }, [design.imports, imports, load]);

  // load any imports we don't have yet
  React.useEffect(() => {
    if (!load) {
      loadImports(imports, f => {
        const nextImports = f(imports);
        setImports(nextImports);
      });
    }
  }, [imports, load]);

  // update theme
  React.useEffect(() => {
    if (!load) loadTheme(design.theme, setTheme);
  }, [design.theme, load]);

  // store selected
  React.useEffect(() => {
    localStorage.setItem('selected', JSON.stringify(selected));
  }, [selected]);

  // add to changes, if needed
  React.useEffect(() => {
    if (!load) {
      // do this stuff lazily to ride out typing sprees
      const timer = setTimeout(() => {
        // If we already have this design object, we must be doing an undo or
        // redo, and therefore no need to add a change
        if (!changes.some(c => c.design === design)) {
          let nextChanges;
          nextChanges = [...changes];
          nextChanges = nextChanges.slice(changeIndex, 10);
          nextChanges.unshift({ design, selected });
          setChanges(nextChanges);
          setChangeIndex(0);
        }
      }, 1000);
      return () => clearTimeout(timer);
    }
  }, [changes, changeIndex, design, load, selected]);

  // persist preview state when it changes
  React.useEffect(() => {
    localStorage.setItem('preview', JSON.stringify(preview));
    // trigger resize so rendered elements can respond accordingly
    window.dispatchEvent(new Event('resize'));
  }, [preview]);

  // set body editing class when preview changes, so we can use some
  // custom CSS for Layer. Yes, this is a bit of a hack :(
  React.useEffect(() => {
    if (preview) {
      document.body.classList.remove('editing');
    } else {
      document.body.classList.add('editing');
    }
  }, [preview]);

  const onKey = React.useCallback(
    event => {
      if (event.metaKey || event.ctrlKey) {
        if (event.key === 'e' || event.key === 'E') {
          event.preventDefault();
          setPreview(!preview);
        } else if (event.key === 'p' && event.shiftKey) {
          const stored = localStorage.getItem('identity');
          if (stored) {
            const identity = JSON.parse(stored);
            publish({
              design,
              ...identity,
              onChange: setDesign,
              onError: error => console.error(error),
            });
          } else {
            console.warn('You need to have published to be able to re-publish');
          }
        }
      }
    },
    [design, preview],
  );

  const onUndo = React.useCallback(() => {
    const nextChangeIndex = Math.min(changeIndex + 1, changes.length - 1);
    const { design: nextDesign, selected: nextSelected } = changes[
      nextChangeIndex
    ];
    setDesign(nextDesign);
    setSelected(nextSelected);
    setChangeIndex(nextChangeIndex);
  }, [changes, changeIndex]);

  const onRedo = React.useCallback(() => {
    const nextChangeIndex = Math.max(changeIndex - 1, 0);
    const { design: nextDesign, selected: nextSelected } = changes[
      nextChangeIndex
    ];
    setDesign(nextDesign);
    setSelected(nextSelected);
    setChangeIndex(nextChangeIndex);
  }, [changes, changeIndex]);

  return (
    <Grommet
      full
      theme={designerTheme}
      themeMode={colorMode}
      dir={rtl ? 'rtl' : undefined}
    >
      <Keyboard target="document" onKeyDown={onKey}>
        <Grid
          fill
          columns={
            responsive === 'small' || preview
              ? 'flex'
              : [
                  ['small', 'medium'],
                  ['1/2', 'flex'],
                  ['small', 'medium'],
                ]
          }
        >
          {responsive !== 'small' && !preview && (
            <Tree
              design={design}
              imports={imports}
              rtl={rtl}
              selected={selected}
              theme={theme}
              colorMode={colorMode}
              setColorMode={setColorMode}
              setDesign={setDesign}
              setRTL={setRTL}
              setSelected={setSelected}
              onRedo={changeIndex > 0 && onRedo}
              onUndo={changeIndex < changes.length - 1 && onUndo}
            />
          )}

          <ErrorCatcher>
            <Canvas
              design={design}
              imports={imports}
              selected={selected}
              preview={preview}
              setDesign={setDesign}
              setSelected={setSelected}
              theme={theme}
            />
          </ErrorCatcher>

          {responsive !== 'small' &&
            !preview &&
            (!selectedComponent ? (
              <ScreenDetails
                design={design}
                selected={selected}
                setDesign={setDesign}
                setSelected={setSelected}
              />
            ) : (
              <Properties
                design={design}
                libraries={libraries}
                theme={theme}
                selected={selected}
                component={selectedComponent}
                setDesign={setDesign}
                setSelected={setSelected}
              />
            ))}
        </Grid>
      </Keyboard>
    </Grommet>
  );
};

export default App;
