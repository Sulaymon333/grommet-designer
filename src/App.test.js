import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import App from './App';

// for README
global.fetch = () =>
  Promise.resolve({
    ok: true,
    text: () => Promise.resolve(''),
  });

describe('App', () => {
  test('create', async () => {
    const user = userEvent.setup();
    const { asFragment } = await render(<App />);
    expect(asFragment()).toMatchSnapshot();

    await user.click(screen.getByTitle('start a new design'));
    expect(asFragment()).toMatchSnapshot();

    await user.type(screen.getByLabelText('name'), 'test design');
    expect(asFragment()).toMatchSnapshot();

    await user.click(screen.getByTitle('create design'));
    expect(asFragment()).toMatchSnapshot();

    await user.click(screen.getByTitle('select screen template'));
    expect(asFragment()).toMatchSnapshot();

    // await user.click(screen.getByText('PageContent'));
    // expect(asFragment()).toMatchSnapshot();

    // await user.click(screen.getByText('PageContent'));
    // expect(asFragment()).toMatchSnapshot();

    // await user.click(screen.getByTitle('add a component'));
    // expect(asFragment()).toMatchSnapshot();
  });
});
