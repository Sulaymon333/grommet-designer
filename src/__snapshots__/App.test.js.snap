// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App empty 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 gguoEx"
  style="height: 100%;"
>
  <div
    class="StyledBox-sc-13pk1d4-0 koRJcD"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 bHvswQ"
    >
      <h1
        class="StyledHeading-sc-1rdh4aw-0 fJrTWC"
      >
        grommet designer
      </h1>
      <p
        class="StyledParagraph-sc-tbetod-0 cJUZhY"
      >
        design with grommet components
      </p>
      <a
        class="StyledButton-sc-323bzc-0 bFQnfs"
        href="/_new"
        title="create a new design"
      >
        Create
      </a>
      <div
        class="StyledBox-sc-13pk1d4-0 jcXjfc"
      />
      <div
        class="StyledBox-sc-13pk1d4-0 iPbgnM FileInput__ContentsBox-sc-1jzq7im-0 iseSTi"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 bznjRF"
        >
          <span
            class="StyledText-sc-1sadyjn-0 fvItiE FileInput__Message-sc-1jzq7im-2 hzOUsx"
          >
            Import - drop design file here
          </span>
          <a
            class="StyledAnchor-sc-1rp7lwl-0 bhYTLn"
            tabindex="0"
          >
            browse
          </a>
        </div>
        <input
          accept=".json"
          class="StyledFileInput-sc-rl7ywv-0 dkCaZm"
          type="file"
        />
      </div>
      <div
        class="StyledBox-sc-13pk1d4-0 extusd"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 hiGxse"
          id="themeMode"
          role="radiogroup"
        >
          <label
            class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 erduql"
            for="themeMode-dark"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 hIYvga StyledRadioButton-sc-g1f6ld-5 chIzBU"
            >
              <input
                checked=""
                class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cQcVYv"
                id="themeMode-dark"
                name="themeMode"
                tabindex="0"
                type="radio"
                value="dark"
              />
              <div
                class="StyledBox-sc-13pk1d4-0 jeRlcU StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 enovQq"
              >
                <svg
                  class="StyledRadioButton__StyledRadioButtonIcon-sc-g1f6ld-3 jmrcKX"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="6"
                  />
                </svg>
              </div>
            </div>
            <span
              class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2"
            >
              dark
            </span>
          </label>
          <div
            class="StyledBox__StyledBoxGap-sc-13pk1d4-1 fuTsyj"
          />
          <label
            class="StyledRadioButton__StyledRadioButtonContainer-sc-g1f6ld-0 erduql"
            for="themeMode-light"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 hIYvga StyledRadioButton-sc-g1f6ld-5 chIzBU"
            >
              <input
                class="StyledRadioButton__StyledRadioButtonInput-sc-g1f6ld-1 cQcVYv"
                id="themeMode-light"
                name="themeMode"
                tabindex="-1"
                type="radio"
                value="light"
              />
              <div
                class="StyledBox-sc-13pk1d4-0 jIEhEE StyledRadioButton__StyledRadioButtonBox-sc-g1f6ld-4 enovQq"
              />
            </div>
            <span
              class="StyledRadioButton__StyledRadioButtonLabel-sc-g1f6ld-2"
            >
              light
            </span>
          </label>
        </div>
        <div
          class="StyledBox__StyledBoxGap-sc-13pk1d4-1 fuTsyj"
        />
        <label
          class="StyledCheckBox__StyledCheckBoxContainer-sc-1dbk5ju-1 llZSKA"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 iiLUML StyledCheckBox-sc-1dbk5ju-6 bXiBYz"
          >
            <input
              class="StyledCheckBox__StyledCheckBoxInput-sc-1dbk5ju-2 hYGeTt"
              type="checkbox"
            />
            <div
              class="StyledBox-sc-13pk1d4-0 dgGvUg StyledCheckBox__StyledCheckBoxBox-sc-1dbk5ju-3 WtXtt"
            />
          </div>
          <span>
            right to left
          </span>
        </label>
      </div>
    </div>
    <div
      class="StyledBox__StyledBoxGap-sc-13pk1d4-1 esHjNp"
    />
    <div
      class="StyledBox-sc-13pk1d4-0 gxyant"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 dxtVkg"
      >
        <header
          class="StyledBox-sc-13pk1d4-0 eBclet"
        >
          <h2
            class="StyledHeading-sc-1rdh4aw-0 bpWVVC"
          >
            tutorials
          </h2>
        </header>
        <a
          class="StyledButton-sc-323bzc-0 kVhqUm"
          href="https://us-central1-grommet-designer.cloudfunctions.net/images/eric-soderberg-hpe-com/designer%20introduction%202a.mp4"
          target="_blank"
        >
          <div
            class="StyledBox-sc-13pk1d4-0 kSZjsO"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 gTflAu"
            >
              <img
                class="StyledImage-sc-ey4zx9-0 fEOYzu"
                src="https://us-central1-grommet-designer.cloudfunctions.net/images/eric-soderberg-hpe-com/designer-tutorial-introduction.png"
              />
            </div>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyJvBD"
            />
            <span
              class="StyledText-sc-1sadyjn-0 ieEvcg"
            >
              introduction
            </span>
          </div>
        </a>
      </div>
      <div
        class="StyledBox__StyledBoxGap-sc-13pk1d4-1 DUqM"
      />
      <div
        class="StyledBox-sc-13pk1d4-0 kUDZbv"
      >
        <h2
          class="StyledHeading-sc-1rdh4aw-0 bpWVVC"
        >
          inspiration
        </h2>
        <div
          class="StyledBox-sc-13pk1d4-0 diGgTQ"
        >
          <a
            class="StyledButton-sc-323bzc-0 kVhqUm"
            href="https://designer.grommet.io/?id=HPE-design-system-hpedesignsystem-hpe-com&mode=edit"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 kSZjsO"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 gTflAu"
              />
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyJvBD"
              />
              <span
                class="StyledText-sc-1sadyjn-0 ieEvcg"
              >
                HPE design system
              </span>
            </div>
          </a>
          <div
            class="StyledBox__StyledBoxGap-sc-13pk1d4-1 fuTsyj"
          />
          <a
            class="StyledButton-sc-323bzc-0 kVhqUm"
            href="https://designer.grommet.io/?id=Card-eric-soderberg-hpe-com&mode=edit"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 kSZjsO"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 gTflAu"
              />
              <div
                class="StyledBox__StyledBoxGap-sc-13pk1d4-1 hyJvBD"
              />
              <span
                class="StyledText-sc-1sadyjn-0 ieEvcg"
              >
                Card
              </span>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`App empty 2`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 gguoEx"
  style="height: 100%;"
>
  <div
    class="StyledGrid-sc-1wofa1l-0 BBBOU"
  >
    <div
      class="StyledGrommet-sc-19lkkz7-0 iZdjoZ"
      id="designer-canvas"
      style="height: 100%; max-width: 100vw;"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 tFPNe"
        draggable="true"
        style="outline: 1px dashed red;"
        tabindex="-1"
      >
        <div
          class="sc-gsTEea hduLXy"
        >
          <p
            class="StyledParagraph-sc-tbetod-0 kWaqig"
          >
            This Box is currently empty. Add components to it, so it can do its layout thing.
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
